lib/hanami/action/params.rb

```diff
@@ -148,6 +148,40 @@ def self.params(&blk)
         validations(&blk || -> {})
       end
 
+      # Define a validation rule.
+      #
+      # @param blk [Proc] the rule definition
+      #
+      # @since 2.x.x
+      #
+      # @example
+      #   class Event < Hanami::Action
+      #     params do
+      #       required(:start_date).value(:date)
+      #       required(:end_date).value(:date)
+      #     end
+      #
+      #     # Rules must be defined after the params schema.
+      #     rule(:start_date, :end_date) do
+      #       if start_date > end_date
+      #         base.failure('event cannot end before it starts')
+      #       end
+      #     end
+      #
+      #     def handle(req, *)
+      #       halt 400 unless req.params.valid?
+      #       # ...
+      #     end
+      #   end
+      def self.rule(*keys, &blk)
+        unless self._validator
+          raise ArgumentError.new(
+            "The params schema must be defined prior to adding rules."
+          )
+        end
```

Comment:
I'm not sure if this guard is necessary, or if the one below is sufficient.

Reply:
@danhealy in my experimentation this one is not required as I see the error even when I remove it as it comes from `validatable.rb` instead:

<img width="621" alt="image" src="https://github.com/hanami/controller/assets/10128/a340a79f-a743-4079-b00c-f0e24ffdae3f">